liver_imaging_analysis/engine/engine.py

```diff
@@ -428,62 +536,166 @@ def test(self, dataloader = None, callback = False):
+            cp path will be loaded from config.
+        epochs : int
+            number of training epochs.
+            If not defined, epochs will be loaded from config.
+        evaluate_epochs : int
+            The number of epochs to evaluate model after. Default is 1.
+        batch_callback_epochs : int
+            The frequency at which per_batch_callback will be called.
+            Expects a number of epochs. Default is 100.
+        save_weight : bool
+            whether to save weights or not. Default is True.
+        save_path : str
+            the path to save weights at if save_weights is True.
+            If not defined, the potential cp path will be loaded
+            from config.
+        test_batch_callback : bool
+            whether to call per_batch_callback during testing or not.
+            Default is False
+
+        Returns
+        -------
+        float
+            the averaged loss calculated during validation
+
+        """
+
+        if automate == True:
+
+            config.network_parameters["num_res_units"] = trial.suggest_int(
+                "res_units_l{}", 2, 5
+            )
+            config.training["optimizer"] = trial.suggest_categorical(
+                "optimizer", ["Adam", "SGD"]
+            )
+            config.training["optimizer_parameters"]["lr"] = trial.suggest_float(
+                "lr", 1e-5, 1e-1, log=True
+            )
+            config.training["loss_name"] = trial.suggest_categorical(
+                "loss_name", ["monai_dice", "monai_general_dice"]
+            )
+
+        tracker = ExperimentTracking(
+            config.name["experiment_name"], config.name["run_name"]
+        )
+        summary_writer = tracker.tb_logger()
+        # if pretrained, will continue on previous checkpoints and previous ClearML task
+        if pretrained:
+            self.load_checkpoint(cp_path)
+            task = tracker.update_clearml_logger()
+            offset = task.get_last_iteration() + 1
+        else:
+            task = tracker.new_clearml_logger()
+            offset = 0
+
+        # add all configs to ClearMl
+        task.connect_configuration(config.__dict__, name="configs")
+        # #add hyperparameters to ClearMl
+        hparams = self.get_hparams(config.network_name)
+        summary_writer.add_hparams(hparams, metric_dict={})
+
+        self.fit(
+            summary_writer=summary_writer,
+            offset=offset,
+            epochs=epochs,
+            evaluate_epochs=evaluate_epochs,
+            batch_callback_epochs=batch_callback_epochs,
+            save_weight=save_weight,
+            save_path=save_path,
+        )
+        # Evaluate on latest saved check point
+        self.load_checkpoint(save_path)
+        final_loss, final_metric = self.test(
+            self.test_dataloader, callback=test_batch_callback
+        )
+        print(
+            "Final test loss:",
+            final_loss,
+        )
+
+        return final_loss
```

Comment:
What is this method supposed to return? losses? or optimized parameters?

Reply:
it performs it its optimization based on the return value